[AMDGPU] Strengthen preload intrinsics to noundef and nonnull

krzysz00 opened the pull request:
The various preloaded registers (workitem IDs, workgroup IDs, and various implicit pointers) always have a finite, invariant, well-defined value throughout a well-defined program.

In cases where the compiler infers or the user declares that some implicit input will not be used (ex. via amdgcn-no-workitem-id-y), the behavior of the entire program is undefined, since that misdeclaration can cause arbitrary other preloaded-register intrinsics to access the wrong register. This case is not expected to arise in practice, but could occur when the no implicit argument attributes were not cleared correctly in the presence of external functions, indrect calls, or other means of executing un-analyzable code. Failure to detect that case would be a bug in the attributor.

This commit updates the documentation to reflect this long-standing reality.

Then, on the basis that all implicit arguments are defined in all correct programs, the intrinsics that return those values are annototated with `noundef``. Some implicit pointer arguments gain a `nonnull`, but the kernel argument segment pointer or implicit argument pointers don't necessarily have this property.

This will prevent spurious calls to `freeze` in front-end optimizations that destroy user-provided ranges on built-in IDs.

(While I'm here, this commit adds a test for `noundef` on kernel arguments which is currently unimplemented)

krzysz00 committed:
[AMDGPU] Strengthen preload intrinsics to noundef and nonnull

The various preloaded registers (workitem IDs, workgroup IDs, various
implicit pointers) always have a finite, inviriant, well-defined value
throughout a well-defined program.

In cases where the compiler infers or the user declares that some
implicit input will not be used (ex. via amdgcn-no-workitem-id-y), the
behavior of the entire program is undefined, since that misdeclaration
can cause arbitrary other preloaded-register intrinsics to access the
wrong register. This case is not expected to arise in practice, but
could occur when the no implicit argument attributes were not cleared
correctly in the prenence of external functions, indrect calls, or
other means of executing un-analyzable code. Failure to detect that
case would be a bug in the attributor.

This commit updates the documentation to reflect this long-standing
reality.

Then, on the basis that all implicit arguments are defined in all
correct programs, the intrinsics that return those values are
annototated with `noundef` and, in the case of implicit pointers,
`nonnull`.

This will prevent spurious calls to `freeze` in front-end
optimizations that destroy user-provided ranges on built-in IDs.

(While I'm here, this commit adds a test for `noundef` on kernel
arguments which is currently unimplemented)

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -12,11 +12,16 @@
 
 def global_ptr_ty : LLVMQualPointerType<1>;
 
+// The amdgpu-no-* attributes (ex amdgpu-no-workitem-id-z) typically inferred
+// by the backend cause whole-program undefined behavior when violated, such as
+// by causing all other preload register intrinsics to return arbitrarily incorrect
+// values. Outside of such IR-level UB, these preloaded registers are always set
```
This logic only really applies in the entry functions

krzysz00 replied:
True. That being said, if I want workitem ID Y and I'm being called from a amdgpu-no-workitem-id-y function, that function has to invoke undefined behavior in order to carry out the function call ... or the calling convention preserves the preloaded registers and those become incorrectly laid out like in the entry function.

I'll probably want a better way to word this

arsenm commented on llvm/test/CodeGen/AMDGPU/lower-kernargs.ll:
```diff
@@ -1116,10 +1116,29 @@ define amdgpu_kernel void @kern_noalias_global_ptr_x2(ptr addrspace(1) noalias %
+; MESA-NEXT:    store volatile ptr addrspace(1) [[PTR_LOAD]], ptr addrspace(1) undef, align 8
+; MESA-NEXT:    ret void
+;
+  store volatile ptr addrspace(1) %ptr, ptr addrspace(1) undef
```
Avoid undef in tests

krzysz00 replied:
Was following existing tests, I can swap this for a poison

arsenm replied:
a null would be better

arsenm reviewed (commented):
We do lower the kernarg intrinsic as null in some cases:

https://github.com/llvm/llvm-project/blob/a91d5c07f2357f10a5378bb3b0e439847f2b8e00/llvm/lib/Target/AMDGPU/SIISelLowering.cpp#L2153


While you're touching this, we should also create a way to declare the new range attribute on intrinsics and set them to the maximum.

krzysz00 commented:
@arsenm Good to know about the kernarg pointer being potentially null, but also, do you have information about when those cases arise? (I suppose in the case where there are no arguments the kernarg pointer is null, but ... why would someone be accessing the kernarg pointer when there are no arguments there?)

Re the ranges, I was going to either do or file a ticket about updating `AMDGPULowerKernelMetadata` to apply `!range` up in the frontend. And now that I've taken a look around for this new range attribute ... I assume what you're wanting is the conservative, calling-convention-independent upper bound in the intrinsic definition?

arsenm commented:
> @arsenm Good to know about the kernarg pointer being potentially null, but also, do you have information about when those cases arise? (I suppose in the case where there are no arguments the kernarg pointer is null, but

Yes 

>  ... why would someone be accessing the kernarg pointer when there are no arguments there?)

Because you don't know the kernel you are being called from in an arbitrary function. You would only know that if you are inlined into the kernel, so some leaf function would have to see null. We redefined it at some point to only allow this in kernels, and use it as an implementation detail in AMDGPULowerKernelArguments. User code is only allowed to use the implicitarg.ptr, which has the same issue. I see there's a test for the 0 sized implicitarg.ptr case returning null, but I don't actually see where that's handled right now.

So we're still left with the question if we can and should support library code that can null check if implicit args are available or not 



> Re the ranges, I was going to either do or file a ticket about updating `AMDGPULowerKernelMetadata` to apply `!range` up in the frontend. And now that I've taken a look around for this new range attribute ... I assume what you're wanting is the conservative, calling-convention-independent upper bound in the intrinsic definition?

Yes. Also, the more specific range metadata can now be replaced with the callsite attribute

arsenm commented:
> Ok, so, implicit argument pointer can reasonably be null, and I'd think library code could reasonably check for implicit arguments being/not being there to avoid crashes or something like that.

The usage inference isn't smart enough to figure this out, and the attributes as defined don't give you the power to express this anyway. So I'm not sure it's worth trying to maintain this.

I do think it's a good idea to split the kernarg and implicitarg cases into a separate PR though. The other cases feel more straightforward

krzysz00 committed:
Don't targ kernarg and implicitarg nonnull, update comments

(Kernarg and implicitarg have complexities I didn't know of, so move
them to a different PR.)

arsenm reviewed (approved):
LGTM with nit. Also, we seem to be missing a way to declare DereferenecableOrNull

krzysz00 committed:
Fix nits
